webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,586 @@
+  } catch (e) {
+    delete req.session.challenge;
+    res.status(400).json({ error: e.message });
+  }
+});
+
+// ----------------------------------------------------------------------------
+// Credential management
+// ----------------------------------------------------------------------------
+
+/**
+ * Return a user and their credentials
+ * (This server only stores one key per username)
+ * Response:
+ * ```{
+ *   username: String,
+ *   credentials: [Credential]
+ * }```
+
+ Credential
+ ```
+ {
+   credId: String,
+   publicKey: String,
+   aaguid: ??,
+   prevCounter: Int
+ };
+ ```
+ **/
+router.post('/getCredentials', csrfCheck, fullSessionCheck, (req, res) => {
+  const user = findUserByUsername(req.session.username);
+  res.status(200).json(user || {});
+});
+
+/**
+ * Remove a credential id attached to the user
+ * Response: empty JSON `{}`
+ **/
+router.post('/removeCredential', csrfCheck, fullSessionCheck, (req, res) => {
+  const { credId } = req.query;
+  const { username } = req.session;
+  const user = findUserByUsername(username);
+  const updatedCredentials = user.credentials.filter((cred) => {
+    return cred.credId !== credId;
+  });
+  updateCredentials(username, updatedCredentials);
+  res.status(200).json(user || {});
+});
+
+/**
+ * Update existing credential
+ * Input: {
+     id: String,
+     newName: String
+ * }```
+ * Response: user as JSON
+ **/
+
+router.post(
+  '/renameCredential',
+  csrfCheck,
+  fullSessionCheck,
+  check('credId').escape(),
+  check('name').notEmpty().escape(),
+  (req, res) => {
+    // Validate the input
+    const validationErrors = validationResult(req);
+    if (!validationErrors.isEmpty()) {
+      return res.status(400).json({ error: validationErrors.array() });
+    }
+    // Rename the credential
+    const { credId, name: newName } = req.query;
+    try {
+      const { username } = req.session;
+      const user = findUserByUsername(username);
+      const { credentials } = user;
+      const indexOfCredentialToUpdate = credentials.findIndex(
+        (el) => el.credId === credId
+      );
+      // change the credential name on a copy of the credentials
+      const updatedCredentials = [...credentials];
+      updatedCredentials[indexOfCredentialToUpdate].name = newName;
+      // update the stored credentials with the copy
+      updateCredentials(username, updatedCredentials);
+      res.status(200).json(user);
+    } catch (e) {
+      res.status(400).json({ error: e.message });
+    }
+  }
+);
+
+/**
+ * Register user credential.
+ * Input format:
+ * ```{
+     id: String,
+     type: 'public-key',
+     rawId: String,
+     response: {
+       clientDataJSON: String,
+       attestationObject: String,
+       signature: String,
+       userHandle: String
+     }
+ * }```
+ **/
+router.post(
+  '/registerResponse',
+  csrfCheck,
+  fullSessionCheck,
+  check('credId').escape(),
+  check('name').notEmpty().escape(),
+  check('transports').isIn(VALID_TRANSPORTS),
```

Comment:
@agektmr FYI this is the validation approach I'm using - same lib for both escaping and validation